### ISSUE TITLE
Could not authenticate to gateway

### ISSUE THREAD

**Issue by jprrezende** [adrienverge/openfortivpn#351]
Hi, I am using version 1.7.1 on Debian 9.
I am receiving authentication error, but, I tested using FortiClient SSLVPN 4.0.2333 for Linux and it is working.

I debugged:

```
joaopaulo@pc-jp:~/Downloads/openfortivpn$ sudo openfortivpn -v -c /etc/openfortivpn/config
DEBUG:  Loaded config file "/etc/openfortivpn/config".
VPN account password: 
DEBUG:  Config host = "remote.stategrid.com.br"
DEBUG:  Config realm = ""
DEBUG:  Config port = "10443"
DEBUG:  Config username = "JOAO PAULO RAMOS REZENDE"
DEBUG:  Config password = "********"
DEBUG:  server_addr: xxx.xxx.xxx.xxx
DEBUG:  server_port: 10443
DEBUG:  gateway_addr: xxx.xxx.xxx.xxx
DEBUG:  gateway_port: 10443
DEBUG:  Gateway certificate validation failed.
DEBUG:  Gateway certificate digest found in white list.
INFO:   Connected to gateway.
ERROR:  Could not authenticate to gateway (No cookie given).
INFO:   Closed connection to gateway.
DEBUG:  server_addr: xxx.xxx.xxx.xxx
DEBUG:  server_port: 10443
DEBUG:  gateway_addr: xxx.xxx.xxx.xxx
DEBUG:  gateway_port: 10443
DEBUG:  Gateway certificate validation failed.
DEBUG:  Gateway certificate digest found in white list.
INFO:   Logged out.
```
So, is it a bug or I am doing something wrong?
Thanks in advance.

**Comment by mrbaseman** [comment]
Hi,

typical reasons for this error are 
* typo in user name or password
* wrong configuration on the fortigate you are trying to connect to (there seem to be some problems when groups are configured, but clients usually specify users, and in my experience each user should be configured separately in the ssl vpn portal)
* expired certificate if user certificates are in use
* restriction to specific IPs while you are coming from elsewhere
the ssl vpn logs on the fortigate may give a hint, but the logged errors sometimes are also misleading ( at least in fortios 5.2 I have seen "unknown user" in the logs, when the user existed, but tried to login with wrong credentials. 
see also #350, #267 and #49

**Comment by DimitriPapadopoulos** [comment]
@jprrezende Hopefully you were able to run _openfortivpn_. I am closing this ticket for now, please do not hesitate to re-open if needed.

**Comment by mrbaseman** [comment]
just as a hint: in the recently released version 1.8.0 we have improved error messages and the debugging output in this area. The error message gives the hint about incorrect credentials, when launched with `-v` option, the http response code is shown (which may already help in this situation), and with `-v -v` the full http traffic is displayed for debugging.

**Comment by jprrezende** [comment]
Hi, sorry for my delay to answer.

I tried again on Debian 10. openfortivpn version 1.8.1, and the result is the same.
My username, password and certificate are correct. I know because I used FortiClient SSLVPN 4.0.2333 to verify this variables.

I used the -v -v arguments, and here is the output:

DEBUG:  openfortivpn 1.8.1
DEBUG:  Loaded config file "/etc/openfortivpn/config".
DEBUG:  Config host = "xxx"
DEBUG:  Config realm = ""
DEBUG:  Config port = "xxx"
DEBUG:  Config username = "xxxxx"
DEBUG:  Config password = "********"
DEBUG:  Resolving gateway host ip
DEBUG:  Establishing ssl connection
DEBUG:  server_addr: xxxx
DEBUG:  server_port: xxx
DEBUG:  gateway_addr: xxx
DEBUG:  gateway_port: xxx
DEBUG:  Gateway certificate validation failed.
DEBUG:  Gateway certificate digest found in white list.
INFO:   Connected to gateway.
DEBUG:  http_send: 
POST /remote/logincheck HTTP/1.1
Host: xxx.xxx.xxx.xxx:xxx
User-Agent: Mozilla/5.0 SV1
Accept: text/plain
Accept-Encoding: identity
Content-Type: application/x-www-form-urlencoded
Cookie: 
Content-Length: 116

username=xxxx&credential=xxxx%2B%2B&realm=&ajax=1&redir=%2Fremote%2Findex&just_logged_in=1
DEBUG:  http_receive: 
HTTP/1.1 200 OK
Date: Sat, 13 Jul 2019 00:08:29 GMT
Set-Cookie:  SVPNCOOKIE=; path=/; expires=Sat, 13-Jul-2019 00:08:29 GMT; secure; httponly;
Set-Cookie: SVPNNETWORKCOOKIE=; path=/remote/network; expires=Sat, 13-Jul-2019 00:08:29 GMT; secure; httponly
X-UA-Compatible: requiresActiveX=true
Transfer-Encoding: chunked
Content-Type: text/html; charset=utf-8
X-Frame-Options: SAMEORIGIN

e54   
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><title>login</title><meta http-equiv="Pragma" content="no-cache"><meta http-equiv="cache-control" content="no-cache"><meta http-equiv="cache-control" content="must-revalidate"><link href="/sslvpn/css/login.css" rel="stylesheet" type="text/css"><script type="text/javascript">if (top && top.location != window.location) top.location = top.location;if (window.opener && window.opener.top) { window.opener.top.location = window.opener.top.location; self.close(); }</script></head><body class="main"><center><table width="100%" height="100%" align="center" class="container" valign="middle" cellpadding="0" cellspacing="0"><tr valign=middle><td><form action="/remote/logincheck" method="post" name="f" autocomplete="off"><table class="list" cellpadding=10 cellspacing=0 align=center width=400 height=180><tr class="dark"><td colspan=2><b>Please Login</b></td></tr><tr style="font-weight:bold"><td id=err_str style="padding-left:10px"></td><td id="err_val" title="sslvpn_login_permission_denied"></td></tr>


<tr><td width=40% style="padding-left:10px"><b>Name:</b></td><td width=60%><input type=text maxlength=128 name=username id=username></td></tr><tr><td width=40% style="padding-left:10px"><b>Password:</b></td><td width=60%><input type=password autocomplete="off" maxlength=128 name=credential id=credential></td></tr><tr id=token_msg style="display:none;"><td colspan=2 id="token_label" style="text-align: left; font-size: 10px; padding-left: 10px"></td></tr>
<tr id=chal_input style="display:none;"><td id=chal_inputlabel width=40% style="padding-left:10px; font-weight: bold;"></td><td width=60%><input type=password autocomplete="off" maxlength=128 name=credential2 id=credential2></td></tr><tr id=chal_input2 style="display:none;"><td id=chal_inputlabel2 width=40% style="padding-left:10px; font-weight: bold;"></td><td width=60%><input type=password autocomplete="off" maxlength=128 name=credential3 id=credential3></td></tr><tr id=fortitoken style="display:none;"><td width=40% style="padding-left:10px"><b id=codelabel>FortiToken Code:</b></td><td width=60%><input type=password autocomplete="off" maxlength=8 name=code id=code></td></tr><tr id=driftmsg style="display:none;"><td colspan=2 style="padding-left:10px">FortiToken clock drift detected. Please input the next code and continue.</td></tr><tr id=fortitoken2 style="display:none;"><td width=40% style="padding-left:10px"><b id=codelabel2>Next FortiToken Code:</b></td><td width=60%><input type=password autocomplete="off" maxlength=8 name=code2 id=code2></td></tr><tr><td></td><td id=login><input type=button name=login_button id=login_button value="Login" onClick="try_login()" border=0></td></tr></table><input type=hidden name=just_logged_in value=1><input type=hidden name=magic id=magic_id value=""><input type=hidden name=reqid id=reqid_id value="0"><input type=hidden name=grpid id=grpid_id value=""><input type=hidden name=realm id=realm_id value=""><input type=hidden name=redir value="/sslvpn/portal.html"><script type="text/javascript" src="/sslvpn/js/login.js?q=6900"></script><script type="text/javascript" src="/lang/en.js"></script><script>document.onkeydown = key_pressdown;function load_login_strings() {var tmp = document.getElementById("err_str");if (tmp) {tmp.innerHTML = fgt_lang["error"] + ":";tmp = document.getElementById("err_val");tmp.innerHTML = fgt_lang[tmp.getAttribute('title')];}}window.onload = load_login_strings;</script></form></td></tr></table></center></body><script>document.forms[0].username.focus();</script></html>

0


ERROR:  Could not authenticate to gateway. Please check the password, client certificate, etc.
DEBUG:  No cookie given -7
INFO:   Closed connection to gateway.
DEBUG:  server_addr: xxxx
DEBUG:  server_port: xxx
DEBUG:  gateway_addr: xxxxx
DEBUG:  gateway_port: xxxx
DEBUG:  Gateway certificate validation failed.
DEBUG:  Gateway certificate digest found in white list.
DEBUG:  http_send: 
GET /remote/logout HTTP/1.1
Host: xxx.xxx.xxx.xxx:xxx
User-Agent: Mozilla/5.0 SV1
Accept: text/plain
Accept-Encoding: identity
Content-Type: application/x-www-form-urlencoded
Cookie: 
Content-Length: 0


DEBUG:  http_receive: 
HTTP/1.1 302 Found
Date: Sat, 13 Jul 2019 00:08:30 GMT
Set-Cookie:  SVPNCOOKIE=; path=/; expires=Sat, 13-Jul-2019 00:08:30 GMT; secure; httponly;
Set-Cookie: SVPNNETWORKCOOKIE=; path=/remote/network; expires=Sat, 13-Jul-2019 00:08:30 GMT; secure; httponly
Location: /remote/logoutok
Content-Length: 0
Content-Type: text/plain
X-Frame-Options: SAMEORIGIN

**Comment by mrbaseman** [comment]
I've just noticed that it might be a problem with the urlencode function. Maybe we should use + instead of % or not encode it at all... I'll try out some variants.

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman The _url_encode()_ function? Ouch, this has been there "forever". Do the username or password contain on-ASCII characters?

**Comment by mrbaseman** [comment]
well not non-ASCII characters, but non-alphanumeric ones, like '+' or '%' for example. Also an interesting candidate is the space character in the password. It should be encoded somehow in the url query, but do we handle it the same way as forticlient does?

**Comment by DimitriPapadopoulos** [comment]
Indeed there is no special case for '+' or '%' but I thought these were handled by this catch-all 'else':
https://github.com/adrienverge/openfortivpn/blob/cbd77ee/src/http.c#L50
As far as I can see '%' is encoded as '%25' and '+' is encoded as '%2B' as expected. Does the Fortigate device expect something different, non-standard?

**Comment by DimitriPapadopoulos** [comment]
The space ' ' is encoded as '+':
https://github.com/adrienverge/openfortivpn/blob/cbd77ee/src/http.c#L48-L49

**Comment by DimitriPapadopoulos** [comment]
Does the Fortigate device expect something different and non-standard?

**Comment by mrbaseman** [comment]
that was my impression, but in my tests with special characters %, +, and space in the password, at least with local users on FortiOS 5.6 everything worked as expected. 

If the password is wrong I get a response code of  405 Method Not Allowed. Whereas in the case reported here, the response is 200 OK, but still there is no cookie given and a comment 
`<!--sslvpnerrmsg=Permission denied.-->`.  
I'm not sure what this means - perhaps credentials are ok, but the user is not allowed to open a tunnel or he can not access anything behind the tunnel. But this contradicts the experience that the official FortiClient works. 

Maybe a compliance rule is configured that [only specific operating systems](https://help.fortinet.com/fos50hlp/56/Content/FortiOS/fortigate-managing-devices/dev-security-policies.htm) are allowed? But then, the whole [endpoint registration and telemetry topic](https://help.fortinet.com/fos50hlp/54/Content/FortiOS/fortigate-security-profiles-54/FortiClient%20Profiles/ec-reg-over-vpn.htm) comes in

**Comment by jprrezende** [comment]
Hi, my username has space characters and password has no alphanumeric characters.

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman As you write, there is indeed the possibility that this is not related to spaces in the username or non-alphanumerc characters in the password.
@jprrezende Is it feasible to try with a username and a password without spaces or non-alphanumeric characters just for the purpose of testing, or is this beyond your reach?

**Comment by jprrezende** [comment]
@DimitriPapadopoulos I can not do it. This credential is managed by another team of the company. Do you suggest another kind of test that I can provide you more information about the problem?

**Comment by jprrezende** [comment]
I will use Wireshark to see the the request and response using the FortiClient SSLVPN 4.0.2333 and after using openfortivpn and then we can compare both messages. Maybe it helps. I will do it tomorrow. Thanks.

**Comment by amanjeev** [comment]
Sending space as-is works for my case. Maybe you can try that, @jprrezende ? See my PR #457.

**Comment by DimitriPapadopoulos** [comment]
@jprrezende Not sure you can see much with Wireshark as the traffic is encrypted.
@amanjeev Interesting, so it seems encoding spaces ` ` as `+` doesn't work. Could you perhaps try `%20`? Indeed there are [two slightly different encdoings out there](https://en.wikipedia.org/wiki/Percent-encoding#The_application/x-www-form-urlencoded_type).

**Comment by mrbaseman** [comment]
hmm... for me (local user on FortiOS 5.6) a space in the user name encoded as `+` works fine. Maybe we have to distinguish between the different authentication methods here (local, ldap, radius, ...)?

**Comment by mrbaseman** [comment]
%20 and plain space character works, too, for me.

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman Interesting. Perhaps we need to identify an encoding that works in all cases - even if it's not standard - or use specific encodings for specific cases. I really don't know. Is there a way to understand what _FortiClient_ does?

**Comment by amanjeev** [comment]
Yea %20 works for me too. #458 as an example.

**Comment by mrbaseman** [comment]
in my setup %20 like in #458 also works for space characters contained in the password, so this might be the encoding that works in all cases that we are looking for... 
To understand what forticlient does one could try to inspect the communication with wireshark, but you have to break up the ssl encrypted channel for the https communication (not actually the tunnel which is opened later), but the authentication part... probably you need to load the certificate and key of the fortigate into wireshark to do this investigation as a man in the middle :) I can do that when I find the time to do so.
Another approach would be to find many people who would like to test #458 and who have space characters in their user name or password or who are able to change the settings for testing purposes

**Comment by amanjeev** [comment]
I agree with @mrbaseman. It is tough and time consuming to try and see what the client does, IMO and I for sure cannot do this. I merely tried this solution and it worked for me. I also peeked into the API call that the webclient makes in our installation and got the encoding idea from there.

**Comment by mrbaseman** [comment]
We have found that it might be better to encode the space character as `%20` instead of `+` now. The next question is how shall a `+` be encoded? as plain `+` or as `%2B`. I can't test this easily right now. I hope that I can do a quick check on this question beginning of next week.

**Comment by amanjeev** [comment]
I wonder if it makes sense to have a config based escape setting that users can choose before using?

**Comment by mrbaseman** [comment]
Hmm... it might make sense if it turns out that the encoding used by Fortinet has changed between two versions (Client or FortiOS) or if it depends on the type of users. However, I believe we just have to find out how to do it correctly to cover all corner-cases.

**Comment by mrbaseman** [comment]
I have done [that test](https://github.com/adrienverge/openfortivpn/issues/351#issuecomment-514788706) now and it turns out that `+` has to be encoded as `%2B`. Changing this and sending the character as plain `+` breaks _openfortivpn_. So, I think #458 is fine (unless we get reports about encoding the space character breaks things somewhere else)

**Comment by mrbaseman** [comment]
we have merged #458 now, which hopefully solves this issue

**Comment by Saratoga6** [comment]
Wonder if  anyone  can help me  down right path to t-shoot this issue.
We all have MAC OSX trying to use SSL VPN connection to Fortigate 500E integrated to AD , via user/password on custom port 10443 , which we know works over the Forticlient  GUI
Not even getting to a point we enter passwords.
-vv  does not tell much more.
Security / FW  stuff note my forte  so any advice be appreciated.
-finn


Client:
ProductName:	Mac OS X
ProductVersion:	10.15.3
BuildVersion:	19D76

% sudo openfortivpn vpn.mydomain.com:10443 -u myemail@mydomain.com-v
DEBUG:  openfortivpn 1.11.0
WARN:   Bad port in config file: "0".
DEBUG:  Loaded config file "/usr/local/Cellar/openfortivpn/1.11.0/etc/openfortivpn/config".
DEBUG:  Config host = "vpn.mydomain.com"
DEBUG:  Config realm = ""
DEBUG:  Config port = "10443"
DEBUG:  Config username = "myemail@mydomain.com"
DEBUG:  Resolving gateway host ip
DEBUG:  Establishing ssl connection
DEBUG:  server_addr: <VPN perimeter IP Fortigate500E>
DEBUG:  server_port: 10443
DEBUG:  gateway_addr: <VPN perimeter IP Fortigate500E>
DEBUG:  gateway_port: 10443
DEBUG:  Setting cipher list to: HIGH:!aNULL:!kRSA:!PSK:!SRP:!MD5:!RC4DEBUG:  Gateway certificate validation succeeded.
INFO:   Connected to gateway.
ERROR:  Could not authenticate to gateway. Please check the password, client certificate, etc.
DEBUG:  HTTP status code 405
INFO:   Closed connection to gateway.
DEBUG:  server_addr: <VPN perimeter IP Fortigate500E>
DEBUG:  server_port: 10443
DEBUG:  gateway_addr: <VPN perimeter IP Fortigate500E>
DEBUG:  gateway_port: 10443
DEBUG:  Setting cipher list to: HIGH:!aNULL:!kRSA:!PSK:!SRP:!MD5:!RC4DEBUG:  Gateway certificate validation succeeded.
INFO:   Logged out.

**Comment by DimitriPapadopoulos** [comment]
The password is incorrect:
```
ERROR: Could not authenticate to gateway. Please check the password, client certificate, etc.
```
That's not a surprise since you don't get to enter enter a password.

The problem here is that _openfortivpn_ reads an incorrect password from its default configuration file and uses that empty password instead of prompting you for the real password. This is a bug specific to version 1.11.0 that has been fixed in 1.12.0.

In 1.11.0 you need to fix this manually. Edit file `/usr/local/Cellar/openfortivpn/1.11.0/etc/openfortivpn/config` and either move it elsewhere or add a leading `# ` to comment out each line of of the file. More precisely it should be enough to change this line:
```
password =
```
into:
```
# password =
```

@mrbaseman Have we reached to the the macOS packagers and asked them to prepare a 1.12.0 release?

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman 1.12.0 is under way on mac0S/Homebrew but it looks like we need to patch:
https://github.com/Homebrew/homebrew-core/pull/50728

**Comment by Saratoga6** [comment]
Brilliant, thank you.  commenting the conf got me past this problem.
I have another , will make separate issue.
-finn

**Comment by DimitriPapadopoulos** [comment]
Change #458 is included in _openfortivpn_ 1.12.0.

**Comment by andy-at-smith** [comment]
I'm running 1.12.0 and have this issue.  The log is basically identical to Saratoga6's above.  With a "password =" line in the config it doesn't work, but when that line is commented out I can enter a password interactively.

This is a Ubuntu 20.04 box, installed via apt:
amason@amason-p53:~$ apt info openfortivpn
Package: openfortivpn
Version: 1.12.0-1
Priority: optional
Section: universe/net

I'm pretty sure the only recent change was an upgrade from 19.10, where it was working.

**Comment by DimitriPapadopoulos** [comment]
@andy-at-smith This issue has been fixed in 1.12.0. This issue is closed. Please create a new issue.

**Comment by andy-at-smith** [comment]
Apparently not, but fine, if I get motivated enough I'll create an issue for the exact same problem, with the exact same logs, for which the exact same resolution worked.  Meh, never mind, I've reported it here, if you feel like fixing it, great, if not, I'm all set.

**Comment by DimitriPapadopoulos** [comment]
@andy-at-smith Perhaps it's the same symptom, but it's certainly not the same issue since the cause has been fixed in 1.12.0.

I suspect your own issue is related to #507.

If this doesn't help, please open a new issue as this one is certainly unrelated.

**Comment by DimitriPapadopoulos** [comment]
Note that on Ubuntu 20.04 the default config file `/etc/openfortivpn/config` looks like this:
```
# ### config file for openfortivpn, see man openfortivpn(1) ###
# 
# host = vpn.example.org
# port = 443
# username = vpnuser
# password = VPNpassw0rd
```
As you can see `/etc/openfortivpn/config` does not contain an empty password. If you have not modified the config file to specify an empty password, then you're experiencing expected behaviour. if you have not modified the config file, then I suspect an existing pre-1.11.0 configuration file has been left over from a previous version of _openfortivpn_.
 
The slight incompatibility between pre-1.11.0 default configuration files and post-1.11.0 is unfortunate but cannot be fixed, as explained in #507. It's too late now and the issue will disappear with time as old incompatible config files disappear or are overwritten or fixed.

The only thing that can be fixed is the _openfortivpn_ upgrade process and how it handles config files. Since you're using an Ubuntu DEB package, upgrading/installing _openfortivpn_ is handled by the package itself, so you should report this issue to Debian/Ubuntu packagers if you want this fixed.

**Comment by mrbaseman** [comment]
@andy-at-smith yes, you are seeing the same issue on Ubuntu as described here for MacOS X/Homebrew. You have a preexisting config file with an empty password, which was ignored before 1.11.0. The fix in 1.12.0 was that the example config file is installed with the password line commented out. However, the default config file is not overwritten. If the file already exists, it is preserved with its current content. We can't edit the file during update and comment out this line, because it has a new meaning starting with 1.11.0, i.e. it allows login without password (e.g. for smartcard support or logging in with a user certificate only). 

The solution to this is - as you write - commenting out the preexisting line with the empty password. That's a manual change of the config file, which is required in conjunction with the update. I'm sorry for that, but that's the least possible pain for now. Changing the behavior again will break the config file once more for other users who rely on the new feature or for those who have already changed their config file.

**Comment by PululuK** [comment]
Hello guys,

I encountered the same problem, in my case it was the authentication parameters which were incorrect, specifically I had put port 433 instead of 443 in the configuration files.

Thanks
